cloudpickle/cloudpickle_fast.py

```diff
@@ -0,0 +1,605 @@
+    if obj is sys.stdin:
+        raise pickle.PicklingError("Cannot pickle standard input")
+    if obj.closed:
+        raise pickle.PicklingError("Cannot pickle closed files")
+    if hasattr(obj, "isatty") and obj.isatty():
+        raise pickle.PicklingError(
+            "Cannot pickle files that map to tty objects"
+        )
+    if "r" not in obj.mode and "+" not in obj.mode:
+        raise pickle.PicklingError(
+            "Cannot pickle files that are not opened for reading: %s"
+            % obj.mode
+        )
+
+    name = obj.name
+
+    retval = io.StringIO()
+
+    try:
+        # Read the whole file
+        curloc = obj.tell()
+        obj.seek(0)
+        contents = obj.read()
+        obj.seek(curloc)
+    except IOError:
+        raise pickle.PicklingError(
+            "Cannot pickle file %s as it cannot be read" % name
+        )
+    retval.write(contents)
+    retval.seek(curloc)
+
+    retval.name = name
+    return _file_reconstructor, (retval,)
+
+
+def _mappingproxy_reduce(obj):
+    return types.MappingProxyType, (dict(obj),)
+
+
+def _memoryview_reduce(obj):
+    return bytes, (obj.tobytes(),)
+
+
+def _module_reduce(obj):
+    if _is_dynamic(obj):
+        return _dynamic_module_reconstructor, (obj.__name__, vars(obj))
+    else:
+        return _module_reconstructor, (obj.__name__,)
+
+
+def _method_reduce(obj):
+    return (types.MethodType, (obj.__func__, obj.__self__))
+
+
+def _logger_reduce(obj):
+    return logging.getLogger, (obj.name,)
+
+
+def _root_logger_reduce(obj):
+    return logging.getLogger, ()
+
+
+def _weakset_reduce(obj):
+    return weakref.WeakSet, (list(obj),)
+
+
+def _dynamic_function_reduce(func, globals_ref):
+    """Reduce a function that is not pickleable via attribute loookup.
+    """
+    newargs = _function_getnewargs(func, globals_ref)
+    state = _function_getstate(func)
+    return types.FunctionType, newargs, state, None, None, _function_setstate
+
+
+def _function_reduce(obj, globals_ref):
+    """Select the reducer depending on obj's dynamic nature
+
+    This functions starts by replicating save_global: trying to retrieve obj
+    from an attribute lookup of a file-backed module. If this check fails, then
+    a custom reducer is called.
+    """
+    if obj in _BUILTIN_TYPE_CONSTRUCTORS:
+        # We keep a special-cased cache of built-in type constructors at
+        # global scope, because these functions are structured very
+        # differently in different python versions and implementations (for
+        # example, they're instances of types.BuiltinFunctionType in
+        # CPython, but they're ordinary types.FunctionType instances in
+        # PyPy).
+        #
+        # If the function we've received is in that cache, we just
+        # serialize it as a lookup into the cache.
+        return _BUILTIN_TYPE_CONSTRUCTORS[obj], ()
+
+    name = obj.__name__
+    try:
+        modname = pickle.whichmodule(obj, name)
+    except Exception:
```

Comment:
Please add a comment detailing the kinds of exceptions we expect and why we ignore them.

Reply:
this specific (retrieving an object from the module it points to) part of cloudpickle implementation is subject to change soon (related to #261)